src/cmd/ksh93/tests/special-dev-paths.sh

```diff
@@ -0,0 +1,159 @@
+[[ -e /dev/stdout ]] || log_error '-e /dev/stdout failed'
+[[ -e /dev/stderr ]] || log_error '-e /dev/stderr failed'
+[[ -e /dev/stdinX ]] && log_error '-e /dev/stdinX succeeded but should have failed'
+[[ -e /dev/stdoutX ]] && log_error '-e /dev/stdoutX succeeded but should have failed'
+[[ -e /dev/stderrX ]] && log_error '-e /dev/stderrX succeeded but should have failed'
+
+[[ -e /dev/fd/0 ]] || log_error '-e /dev/fd/0 failed'
+# TODO: This test currently fails. It thinks the bogus path is valid.
+# [[ -e /dev/fd/xxx ]] && log_error '-e /dev/fd/xxx succeeded but should have failed'
+
+[[ -e /dev/tcp/localhost/666 ]] || log_error '-e /dev/tcp/localhost/666 failed'
+[[ -e /dev/TCP/localhost/666 ]] && \
+    log_error '-e /dev/TCP/localhost/666 succeeded but should have failed'
+# TODO: This test currently fails. It thinks the bogus path is valid.
+# [[ -e /dev/tcp/a_host ]] && \
+#    log_error '-e /dev/TCP/a_host succeeded but should have failed'
+
+[[ -e /dev/udp/localhost/666 ]] || log_error '-e /dev/udp/localhost/666 failed'
+[[ -e /dev/UDP/localhost/666 ]] && \
+    log_error '-e /dev/UDP/localhost/666 succeeded but should have failed'
+# TODO: This test currently fails. It thinks the bogus path is valid.
+# [[ -e /dev/udp/a_host ]] && \
+#     log_error '-e /dev/udp/a_host succeeded but should have failed'
+
+# TODO: This test currently fails. It thinks the bogus path is valid.
+# [[ -e /dev/tcp/ ]] && log_error '-e /dev/tcp/ succeeded but should have failed'
+
+# The tests we just ran should not have opened any fds.
+if [[ $fd_check == yes ]]
+then
+    new_fd_count=$( ls -1 /proc/$$/fd | wc -l )
+    [[ $new_fd_count -eq $fd_count ]] || \
+        log_error 'fds unexpectedly opened' "$fd_count" "$new_fd_count"
+fi
+
+# ==========
+# Verify that opening the special /dev/stdXXX paths works. This does not require the system to
+# natively support these paths. The emulation should also result in a new fd being assigned.
+exec {actual_fd}</dev/stdin
+expect_fd=10
+[[ $actual_fd -ge $expect_fd ]] || \
+    log_error 'Opening /dev/stdin failed' "$expect_fd" "$actual_fd"
+
+exec {actual_fd}>/dev/stdout
+expect_fd=10
+[[ $actual_fd -ge $expect_fd ]] || \
+    log_error 'Opening /dev/stdout failed' "$expect_fd" "$actual_fd"
+
+exec {actual_fd}>/dev/stderr
+expect_fd=10
+[[ $actual_fd -ge $expect_fd ]] || \
+    log_error 'Opening /dev/stderr failed' "$expect_fd" "$actual_fd"
+
+# The tests we just ran should have opened three fds.
+if [[ $fd_check == yes ]]
+then
+    new_fd_count=$( ls -1 /proc/$$/fd | wc -l )
+    (( fd_count += 3 ))
+    [[ new_fd_count -eq fd_count ]] || log_error 'fd count wrong' "$fd_count" "$new_fd_count"
+fi
+
+# ==========
+# Verify that opening the special /dev/tcp/host/service path works.
+# This requires the netcat program.
+if [[ $nc_available == no ]]
+then
+    log_info 'Skipping the /dev/tcp/... I/O tests because nc is not available'
+else
+    nc -l $host $port >nc.out &
```

Comment:
This test can not be executed in parallel. It is failing on Travis with this error:

```
Ncat: bind to 127.0.0.1:65123: Address already in use. QUITTING.
```

Reply:
Doh! Of course. I'll blacklist this one to run serially.